File: knowledge/concepts/pid-controller.c
/* PID Controller -- C implementation
 * First-order plant: y[k+1] = y[k] + (dt/tau) * (u[k] - y[k])
 * PID: u = Kp*e + Ki*I + Kd*D
 * Outputs y at each timestep for verification.
 */
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
    /* Parameters (can be overridden via command line) */
    double Kp = 2.0, Ki = 5.0, Kd = 0.1;
    double tau = 1.0, dt = 0.01, setpoint = 1.0;
    double sim_time = 10.0;

    if (argc >= 7) {
        Kp = atof(argv[1]);
        Ki = atof(argv[2]);
        Kd = atof(argv[3]);
        tau = atof(argv[4]);
        dt = atof(argv[5]);
        setpoint = atof(argv[6]);
    }
    if (argc >= 8) {
        sim_time = atof(argv[7]);
    }

    int steps = (int)(sim_time / dt);
    double y = 0.0;      /* process variable */
    double integral = 0.0;
    double e_prev = 0.0;

    for (int k = 0; k < steps; k++) {
        double e = setpoint - y;
        integral += e * dt;
        double derivative = (e - e_prev) / dt;
        double u = Kp * e + Ki * integral + Kd * derivative;

        /* First-order plant */
        y = y + (dt / tau) * (u - y);

        /* Output: time, y, u */
        printf("%.4f %.8f %.8f\n", k * dt, y, u);

        e_prev = e;
    }

    return 0;
}

File: knowledge/concepts/pid-controller.rb
# PID Controller -- Ruby implementation
# First-order plant: y[k+1] = y[k] + (dt/tau) * (u[k] - y[k])
# PID: u = Kp*e + Ki*I + Kd*D
# Outputs y at each timestep for verification.

def simulate_pid(kp: 2.0, ki: 5.0, kd: 0.1, tau: 1.0, dt: 0.01,
                 setpoint: 1.0, sim_time: 10.0)
  steps = (sim_time / dt).to_i
  y = 0.0
  integral = 0.0
  e_prev = 0.0
  results = []

  steps.times do |k|
    e = setpoint - y
    integral += e * dt
    derivative = (e - e_prev) / dt
    u = kp * e + ki * integral + kd * derivative

    # First-order plant
    y = y + (dt / tau) * (u - y)

    results << { t: k * dt, y: y, u: u }
    e_prev = e
  end

  results
end

if __FILE__ == $0
  results = simulate_pid
  results.each do |r|
    printf("%.4f %.8f %.8f\n", r[:t], r[:y], r[:u])
  end
end